cpu/sam0_common/periph/eth.c

```diff
@@ -0,0 +1,319 @@
+static struct eth_buf_desc tx_desc[ETH_TX_BUFFER_COUNT] __attribute__((aligned(GMAC_DESC_ALIGNMENT)));
+
+static struct eth_buf_desc *rx_curr;
+static struct eth_buf_desc *tx_curr;
+
+/* Declare our own indexes to point to a RX/TX buffer descriptor.
+   GMAC IP have its own indexes on its side */
+static uint8_t  tx_idx;
+static uint8_t  rx_idx;
+
+static uint8_t  rx_buf[ETH_RX_BUFFER_COUNT][ETH_RX_BUFFER_SIZE] __attribute__((aligned(GMAC_BUF_ALIGNMENT)));
+static uint8_t  tx_buf[ETH_TX_BUFFER_COUNT][ETH_TX_BUFFER_SIZE] __attribute__((aligned(GMAC_BUF_ALIGNMENT)));
+
+static void _init_desc_buf(void)
+{
+    int i;
+    /* Initialize RX buffer descriptors */
+    for (i=0; i < ETH_RX_BUFFER_COUNT; i++) {
+        rx_desc[i].address = ((uint32_t) (rx_buf[i]) & DESC_RX_ADDR_ADDR_MASK);
+    }
+    /* Set WRAP flag to indicate last buffer */
+    rx_desc[i-1].address |= DESC_RX_ADDR_WRAP;
+    rx_curr = &rx_desc[0];
+    /* Initialize TX buffer descriptors */
+    for (i=0; i < ETH_TX_BUFFER_COUNT; i++) {
+        tx_desc[i].address = (uint32_t) tx_buf[i];
+    }
+    /* Set WRAP flag to indicate last buffer */
+    tx_desc[i-1].status |= DESC_TX_STATUS_WRAP;
+    tx_curr = &tx_desc[0];
+    /* Setup buffers index */
+    rx_idx = 0;
+    tx_idx = 0;
+    /* Store RX buffer descriptor list */
+    GMAC->RBQB.reg = (uint32_t) rx_desc;
+    /* Store TX buffer descriptor list */
+    GMAC->TBQB.reg = (uint32_t) tx_desc;
+}
+
+int sam0_read_phy(uint8_t phy, uint8_t addr)
+{
+    GMAC->MAN.reg = GMAC_MAN_REGA(addr) | GMAC_MAN_PHYA(phy)
+                  | GMAC_MAN_CLTTO      | GMAC_MAN_WTN(0x2)
+                  | GMAC_MAN_OP(PHY_READ_OP);
+
+    /* Wait for operation completion */
+    while (!GMAC->NSR.bit.IDLE) {}
+    /* return content of shift register */
+    return (GMAC->MAN.reg & GMAC_MAN_DATA_Msk);
+}
+
+void sam0_write_phy(uint8_t phy, uint8_t addr, uint16_t data)
+{
+    GMAC->MAN.reg = GMAC_MAN_REGA(addr) | GMAC_MAN_PHYA(phy)
+                  | GMAC_MAN_WTN(0x2)   | GMAC_MAN_OP(PHY_WRITE_OP)
+                  | GMAC_MAN_CLTTO      | GMAC_MAN_DATA(data);
+
+    /* Wait for operation completion */
+    while (!GMAC->NSR.bit.IDLE) {}
+}
+
+void sam0_eth_set_mac(const eui48_t *mac)
+{
+    GMAC->Sa[0].SAT.reg = ((mac->uint8[5] << 8)  | mac->uint8[4]);
+    GMAC->Sa[0].SAB.reg = ((mac->uint8[3] << 24) | (mac->uint8[2] << 16)
+                        |  (mac->uint8[1] << 8)  | mac->uint8[0]);
+}
+
+void sam0_eth_get_mac(eui48_t *out)
+{
+    out->uint8[5] = (GMAC->Sa[0].SAT.reg >> 8);
+    out->uint8[4] = (GMAC->Sa[0].SAT.reg);
+    out->uint8[3] = (GMAC->Sa[0].SAB.reg >> 24);
+    out->uint8[2] = (GMAC->Sa[0].SAB.reg >> 16);
+    out->uint8[1] = (GMAC->Sa[0].SAB.reg >> 8);
+    out->uint8[0] = (GMAC->Sa[0].SAB.reg);
+}
+
+int sam0_eth_send(const struct iolist *iolist)
+{
+    unsigned len = iolist_size(iolist);
+    unsigned tx_len = 0;
+    tx_curr = &tx_desc[tx_idx];
```

Comment:
What if this `tx_idx` is already in use?

Reply:
I agree that we could add a guard here. If this case occurs, I think increasing the number of allocated TX buffers should be mitigate the problem as it depends on the application ethernet load.